addons/mass_mailing/controllers/main.py

```diff
@@ -102,6 +102,64 @@ def mailing_unsubscribe_oneclick(self, mailing_id, document_id=None, email=None,
         self.mailing_unsubscribe(mailing_id, document_id=document_id, email=email, hash_token=hash_token, **post)
         return Response(status=200)
 
+    @http.route(['/mailing/<int:mailing_id>/confirm_unsubscribe'], type='http', website=True, auth='public')
+    def mailing_confirm_unsubscribe(self, mailing_id, document_id=None, email=None, hash_token=None):
+        mailing = request.env['mailing.mailing'].sudo().browse(mailing_id)
+
+        unsubscribed_list = ''
+        email_found, hash_token_found = self._fetch_user_information(email, hash_token)
+        try:
+            self._check_mailing_email_token(
+                mailing_id, document_id, email_found, hash_token_found,
+                required_mailing_id=True
+            )
+            if mailing.mailing_model_real == 'mailing.contact':
+                unsubscribed_list = ', '.join(str(list.name) for list in mailing.contact_list_ids if list.is_public)
+        except NotFound as e:  # avoid leaking ID existence
+            raise Unauthorized() from e
+        template = request.env.ref('mass_mailing.page_confirm_unsubscribe',raise_if_not_found=False)
+        if template:
+            return request.render('mass_mailing.page_confirm_unsubscribe', {
+                    'mailing_id': mailing_id,
+                    'document_id': document_id,
+                    'email': email,
+                    'hash_token': hash_token,
+                    'unsubscribed_list': unsubscribed_list,
+                    'unsub_confirmed': False,
+                })
+        else:
+            return self.mailing_confirm_unsubscribe_post(mailing_id, document_id, email, hash_token)
+
+    @http.route(['/mailing/confirm_unsubscribe'], type='http', website=True, auth='public', methods=['POST'])
+    def mailing_confirm_unsubscribe_post(self, mailing_id, document_id=None, email=None, hash_token=None, unsubscribed_list=''):
+        # Unsubscribe user
+        email_found, hash_token_found = self._fetch_user_information(email, hash_token)
+        try:
+            mailing_sudo = self._check_mailing_email_token(
+                int(mailing_id), document_id, email_found, hash_token_found,
+                required_mailing_id=True
+            )
+        except NotFound as e:  # avoid leaking ID existence
+            raise Unauthorized() from e
+
+        if mailing_sudo.mailing_on_mailing_list:
+            self._mailing_unsubscribe_from_list(mailing_sudo, document_id, email_found, hash_token_found)
+        else:
+            self._mailing_unsubscribe_from_document(mailing_sudo, document_id, email_found, hash_token_found)
```

Comment:
These methods return renders of the correct templates wouldn't it be possible to use them inside the template so that we don't need to have a redirect inside it ?